File: TP01/palindromo.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define bool short
#define true 1
#define false 0
#define NUMENTRADA 1000
#define TAMLINHA 1000

//Funca para verificar se determinada string e a palavra "FIM"
bool isFim(char *s)
{
	return (strlen(s) >= 3 && s[0] == 'F' && s[1] == 'I' && s[2] == 'M');
}

//Funcao para comparar duas strings, iterando por seus caracteres
bool equals(char *a, char *b)
{
	bool res = true;

	for (int i = 0; i < strlen(b); i++)
	{
		if (a[i] != b[i])
		{
			res = false;
		}
	}

	return res;
}

//Funcao para checar o palindromo
void isPalindromo(char *s)
{
	char buff[strlen(s)];

	//For trocando os caracteres de lugar, entre a string original e o buffer
	for (int i = 0; i <= strlen(s); i++)
	{
		buff[i] = s[strlen(s) - 2 - i];
	}

	if (equals(s, buff))
	{
		printf("SIM\n");
	}
	else
	{
		printf("NAO\n");
	}
}

int main(int argc, char **argv)
{
	char entrada[NUMENTRADA][TAMLINHA];
	int numEntrada = 0;

	//Leitura da entrada padrao
	do
	{
		fgets(entrada[numEntrada], TAMLINHA, stdin);
	} while (isFim(entrada[numEntrada++]) == false);
	numEntrada--; //Desconsiderar ultima linha contendo a palavra FIM

	//Para cada linha de entrada, compara o palindromo
	for (int i = 0; i < numEntrada; i++)
	{
		isPalindromo(entrada[i]);
	}
}

File: TP01/palindromo.java
class palindromo {
   
	//Funcao que troca iterativamente os caracteres de posicao
	public static String reverseString (String s){
		char[] reverse = new char[s.length()];

		for(int i = 0; i < s.length(); i++){
			reverse[i] = s.charAt(s.length() - 1 - i);
		}

	    String result = new String(reverse);

		return result;
	}
     
	//Funcao que remove espacos, iterando por seus caracteres e copiando o que nao for um espaco para uma nova string
	public static String removeSpaces(String s){
		char[] chars = new char[s.length()];

		for (int i = 0; i < s.length(); i++) {
			if(s.charAt(i) != ' '){
				chars[i] = s.charAt(i);
			}
		}

		String result = new String(chars);

		return result;
	}

	//Funcao que compara iterativamente, char por char, duas string
	public static boolean equals(String a, String b){
		boolean result = true;

		if(a.length() != b.length()){
			result = false;
		}else{
			for(int i = 0; i < a.length(); i++){
				if(a.charAt(i) != b.charAt(i)){
					result = false;
					break;
				}
			}
		}

		return result;
	}
     
	//Funcao que verifica um palindromo, utilizando das funcoes anteriores
	public static void isPalindromo(String s){
		String buffer = removeSpaces(s);

		String buff = reverseString(buffer);

		if(equals(buffer, buff)){
			MyIO.println("SIM");
		}else{
			MyIO.println("NAO");
		}
	}

	public static void main (String[] args){
	   String[] entrada = new String[1000];
	   int numEntrada = 0;

	   //Leitura da entrada padrao
	   do {
	      entrada[numEntrada] = MyIO.readLine();
	   } while (equals(entrada[numEntrada++], "FIM") == false);
	   numEntrada--;   //Desconsiderar ultima linha contendo a palavra FIM
     
	   //Para cada linha de entrada, verificando palindromos
	   for(int i = 0; i < numEntrada; i++){
	      isPalindromo(entrada[i]);
	   }
	}
     }
     